File: Data Structures & Algorithms/subsets/submission-1.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> subset; 
        dfs(0,subset, nums, res);
        return res; 
    }

    void dfs(int i, vector<int>& subset, vector<int> &nums, vector<vector<int>>& res) // i = index of the value we are making the decision on 
    {
        if (i == nums.size())
        {
            res.push_back(subset);
            return;
        }
        
        subset.push_back(nums[i]);
        dfs(i+1, subset, nums, res); 

        subset.pop_back(); 
        dfs(i+1, subset, nums, res);
    }
};

File: Data Structures & Algorithms/combinations/submission-1.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res = []

        def backtrack(index, subset):
            if index > n or len(subset) >= k:
                if len(subset) >= k: 
                    res.append(subset[:])
                return 
            
            subset.append(index)
            backtrack(index + 1, subset)
            subset.pop()
            backtrack(index + 1, subset)
        backtrack(1, [])
        return res
